Raise ArgumentError when default has invalid type

v0idpwn opened the pull request:
This is achieved through a dump on the default value. We ignore nil
values and empty lists. Parameterized types use their own dump
implementation.

I added a failing test so we can discuss something: I noticed some
parameterized types will raise on `dump` if the input is not of the
expected type. What shall we do in this scenario? Catch those
errors and build the default argument error?

v0idpwn commented:
Oh, and this closes #3483

josevalim commented on lib/ecto/schema.ex:
```diff
@@ -2061,6 +2062,30 @@ defmodule Ecto.Schema do
+  defp validate_default!(_, nil), do: :ok
+  defp validate_default!(_, []), do: :ok
+
+  defp validate_default!({:parameterized, mod, _} = type, value) do 
```
Do we need this clause? Ecto.Type will already do the same for us, no?

v0idpwn replied:
You're correct! :) 
Removed it.

josevalim commented on lib/ecto/schema.ex:
```diff
@@ -2061,6 +2062,30 @@ defmodule Ecto.Schema do
   end
 
+  defp validate_default!(_, nil), do: :ok
+  defp validate_default!(_, []), do: :ok
```
Do we need these clauses? They should also be handled by Ecto.Type.dump?

v0idpwn replied:
So, I'm in doubt on this one: 

Without the `nil` clause, some custom types will raise on dump. An example is: Ecto.Repo.AutogenerateTest.ParameterizedTypePrefixedID

Without the `[]` clause, I get failures with test scenarios that rely on `embeds_many SchemaDefinedInTestFilesOnly`, as the compiler can't seem to find the module for the Ecto.Embedded parameterized type.

v0idpwn replied:
I believe something could be done in this direction, but I probably need help to find the best approach

josevalim replied:
Then it this case, you either need to pass a default or you need to change `Ecto.Repo.AutogenerateTest.ParameterizedTypePrefixedID` to accept a `nil` on dump. I.e. we are actually catching a bug. 👍 

For embeds, you should probably move the place where the check happens. Check it on `__field__` directly so we don't run it for embeds and the like.

v0idpwn replied:
Managed to remove both of them, had to change some test schemas.

josevalim commented:
I have added some comments. Can you please look at the tests too? Thank you!

v0idpwn commented:
Hi, José!
About the failing test, I added it because I wanted an opinion. Some
parameterized types will raise instead of returning an error in the dump/2
function when receiving invalid input. The failing test reproduces this
scenario. Should I catch those errors?

About the clauses, I'll try to improve them




Em sex, 20 de nov de 2020 05:29, José Valim <notifications@github.com>
escreveu:

> I have added some comments. Can you please look at the tests too? Thank
> you!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-ecto/ecto/pull/3487#issuecomment-731022764>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADSBOVW6M6GWFLCC6OO2EXDSQYSGRANCNFSM4T4DSY6Q>
> .
>

josevalim commented:
I think we can just let them raise for now, yes.

v0idpwn commented:
I believe it is ready now. If there is any other issue, let me know :)

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
